Refactor/hilbert series with flint

Dioprz opened the pull request:
### Description
- Complete reimplementation of the nmonomials class to use python-flint power series instead of Sage power series.
  - Remove of the `_ring` attribute, replacing its use by the new attribute `_gen`.
  - Remove the `series` method, to be replaced by the new methods `coefficient_of_degree` and `coefficient_up_to_degree`.
  - Rename the method `first_nonpositive_integer` as `first_nonpositive_coefficient`.
- Updated all references to the Hilbert Series class to ensure compatibility with the new implementation.
- Stylistic changes via Black formatter.

### Review process
- Major changes are in the reimplementation of HilbertSeries and the used form to call the series in Crossbred.
- Any other change should be minor or stylistic.


### Pre-approval checklist
- [x] The code builds clean without any errors or warnings
- [x] I've added/updated tests
- [x] I've added/updated documentation if needed

Javierverbel reviewed (changes requested):
One test in `cryptographic_estimators/MQEstimator/series/hilbert.py` is failing.

Dioprz commented:
> One test in `cryptographic_estimators/MQEstimator/series/hilbert.py` is failing.

Fixed. I also changed an inappropriate function implementation remaining in `hilbert.py`.

FloydZ commented on cryptographic_estimators/MQEstimator/series/hilbert.py:
```diff
@@ -41,27 +40,59 @@ class HilbertSeries(object):
-        self._ring = PowerSeriesRing(
-            QQ, 'z', default_prec=2*len(self._degrees))
-        z = self._ring.gen()
+        self._prec = 2 * len(self._degrees)
```
why twice the degree?

Javierverbel replied:
The default by PowerSeries is too small. I've noticed that `self._prec = 2 * len(self._degrees)`  should be enough and still fast for most of the cases.

FloydZ replied:
Maybe we should document this? In a comment or something.
Because maybe one day the precision will not be enough (maybe because of big numbers). Or is this completely impossible.

Dioprz replied:
@Javierverbel What do you think? 

Does the parameters have any correlation that forces `n` to grow at the same pace that `degrees`, or we may end up having problems on big `n`'s (or `q`)?

Javierverbel replied:
I would add something like  `This precision should work for most of the case since the systems excepted to be evaluated here have a number of variables which is at most the number of equations`

Dioprz replied:
Perfect, see my last commit.

FloydZ replied:
perfect from my side, can be merged

FloydZ commented on cryptographic_estimators/MQEstimator/series/hilbert.py:
```diff
@@ -102,90 +135,106 @@ def precision(self):
-        return self.ring.default_prec()
+        x = self._gen
+        serie_repr = x.repr()
+        return int(serie_repr[serie_repr.rfind("=") + 1 : serie_repr.rfind(")")])
```
this looks rather error prone. Wouldn't it be easier to store the `default_prec` from the constructor in a separate variable? And the return this variable here?

Dioprz replied:
Fully agree. I don't know why I forgot the 'self._prec' attribute.

Done!

FloydZ replied:
:+1:

FloydZ commented on cryptographic_estimators/MQEstimator/series/hilbert.py:
```diff
@@ -102,90 +135,106 @@ def precision(self):
+            if self._series_up_to_degree[d] <= 0:
+                return int(d)
         raise ValueError(
-            "unable to find a nonpositive coefficient in the series")
```
is it good to directly raise an exception? Im not familiar with Hilbert series, is it common that a series  does not contain a nonpositive coefficient? Or is this a rather rare thing to happen?
If its common, I think its a bad Idea to directly raise an exception. Maybe one could return 0?

FloydZ replied:
BTW: is it a good idea to raise exception in the "hot" code at all?

Javierverbel replied:
You are raising a good point. I'm afraid returning 0 might lead a misinterpretations of the results of the estimator. I think if we set the "Error" case to 0, in order to avoid misinterpretations, we have to check if the output of this method is > than 0 every time we call it, and I think this is worse than raising an error. 

When calling that function the user is expecting a nonzero degree. I think it is better make it aware of if when such a degree cannot be returned.

Javierverbel replied:
We don't expect an MQEstimator user to always to use `try catch`.

FloydZ replied:
ok than its fine be me
